executor/analyze.go

```diff
@@ -381,3 +398,401 @@ func (e *AnalyzeColumnsExec) buildStats() (hists []*statistics.Histogram, cms []
 	}
 	return hists, cms, nil
 }
+
+func analyzeFastExec(exec *AnalyzeFastExec) statistics.AnalyzeResult {
+	hists, cms, err := exec.buildStats()
+	if err != nil {
+		return statistics.AnalyzeResult{Err: err}
+	}
+	result := statistics.AnalyzeResult{
+		PhysicalTableID: exec.PhysicalTableID,
+		Hist:            hists,
+		Cms:             cms,
+	}
+	hist := hists[0]
+	result.Count = hist.NullCount
+	if hist.Len() > 0 {
+		result.Count += hist.Buckets[hist.Len()-1].Count
+	}
+	return result
+}
+
+// AnalyzeFastTask is the task for build stats.
+type AnalyzeFastTask struct {
+	Location  *tikv.KeyLocation
+	isScan    bool
+	SampSize  uint64
+	LRowCount uint64
+	RRowCount uint64
+}
+
+// AnalyzeFastExec represents Fast Analyze Columns executor.
+type AnalyzeFastExec struct {
+	ctx             sessionctx.Context
+	PhysicalTableID int64
+	pkInfo          *model.ColumnInfo
+	colsInfo        []*model.ColumnInfo
+	idxInfo         *model.IndexInfo
+	concurrency     int
+	maxNumBuckets   uint64
+	table           table.Table
+	cache           *tikv.RegionCache
+	wg              *sync.WaitGroup
+	sampLocs        chan *tikv.KeyLocation
+	sampLocRowCount uint64
+	tasks           chan *AnalyzeFastTask
+}
+
+func (e *AnalyzeFastExec) getSampRegionsRowCount(bo *tikv.Backoffer, needRebuild *bool, err *error) {
+	defer func() {
+		e.wg.Done()
+		if *needRebuild == true {
+			for _, ok := <-e.sampLocs; ok; _, ok = <-e.sampLocs {
+				// do nothing
+			}
+		}
+	}()
+	client := e.ctx.GetStore().(tikv.Storage).GetTiKVClient()
+	for {
+		loc, ok := <-e.sampLocs
+		if !ok {
+			return
+		}
+		req := &tikvrpc.Request{
+			Type: tikvrpc.CmdDebugGetRegionProperties,
+			DebugGetRegionProperties: &debugpb.GetRegionPropertiesRequest{
+				RegionId: loc.Region.GetID(),
+			},
+		}
+		var resp *tikvrpc.Response
+		var rpcCtx *tikv.RPCContext
+		rpcCtx, *err = e.cache.GetRPCContext(bo, loc.Region)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+		ctx := context.Background()
+		resp, *err = client.SendRequest(ctx, rpcCtx.Addr, req, tikv.ReadTimeoutMedium)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+		// TODO: check the region not_found
+		if ctx.Err() != nil {
+			fmt.Println(ctx.Err().Error())
+			*needRebuild = true
+			return
+		}
+		for _, prop := range resp.DebugGetRegionProperties.Props {
+			if prop.Name == "num_rows" {
+				var cnt uint64
+				cnt, *err = strconv.ParseUint(prop.Value, 10, 64)
+				if *err != nil {
+					*err = errors.Trace(*err)
+					return
+				}
+				newCount := atomic.AddUint64(&e.sampLocRowCount, cnt)
+				task := &AnalyzeFastTask{
+					Location:  loc,
+					LRowCount: newCount - cnt,
+					RRowCount: newCount,
+				}
+				e.tasks <- task
+			}
+		}
+	}
+}
+
+func (e *AnalyzeFastExec) buildSampTask() (bool, error) {
+	if e.wg == nil {
+		e.wg = &sync.WaitGroup{}
+	}
+	e.wg.Wait()
```

Comment:
Why wait here?

Reply:
Because when meeting rebuild `RegionCache`, the last time goroutine may not end.

Reply:
It should be: 
```suggestion
else {
    e.wg.Wait()
}
```
Otherwise there will be a deadlock in the first call of `buildSampTask`?

Reply:
Oh, never mind, it's ok to wait on a empty `WaitGroup` in `Golang`.

Reply:
Now we don't need it?

Reply:
Now it's useless, but keeping it here has no impact.

Reply:
But it will confuse the reviewers.

Reply:
If we handle the region error, it's necessary to wait here.

Reply:
Why? After this function exits, all the goroutine must also exit.

Reply:
Okay.